src/Authoring/WinRT.SourceGenerator/Generator.cs

```diff
@@ -1,186 +1,263 @@
+            Logger.Log("Running " + cswinrtExe + " " + arguments);
+
+            var processInfo = new ProcessStartInfo
+            {
+                FileName = cswinrtExe,
+                Arguments = arguments,
+                UseShellExecute = false,
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                WindowStyle = ProcessWindowStyle.Hidden,
+                CreateNoWindow = true
+            };
+
+            using var cswinrtProcess = Process.Start(processInfo);
+            Logger.Log(cswinrtProcess.StandardOutput.ReadToEnd());
+            Logger.Log(cswinrtProcess.StandardError.ReadToEnd());
+            cswinrtProcess.WaitForExit();
+
+            foreach(var file in Directory.GetFiles(outputDir, "*.cs", SearchOption.TopDirectoryOnly))
+            {
+                Logger.Log("Adding " + file);
+                context.AddSource(Path.GetFileNameWithoutExtension(file), SourceText.From(File.ReadAllText(file), Encoding.UTF8));
+            }
+
+            // Directory.Delete(outputDir, true);
+        }
+
+        private string GetWinmdOutputFile(GeneratorExecutionContext context)
+        {
+            return Path.Combine(GetGeneratedFilesDir(context), GetAssemblyName(context) + ".winmd");
+        }
+
+        private void GenerateWinMD(MetadataBuilder metadataBuilder, string outputFile) 
+        {
+            Logger.Log("Writing " + outputFile);
+            var managedPeBuilder = new ManagedPEBuilder(
+                new PEHeaderBuilder(
+                    machine: Machine.I386,
+                    imageCharacteristics: Characteristics.ExecutableImage | Characteristics.Dll | Characteristics.Bit32Machine),
+                new MetadataRootBuilder(metadataBuilder, "WindowsRuntime 1.4"),
+                new BlobBuilder(),
+                flags: CorFlags.ILOnly);
+
+            var peBlob = new BlobBuilder();
+            managedPeBuilder.Serialize(peBlob);
+
+            using var fs = new FileStream(outputFile, FileMode.Create, FileAccess.Write);
+            peBlob.WriteContentTo(fs);
+        }
+
+        
+        private bool CatchWinRTDiagnostics(ref GeneratorExecutionContext context) 
+        { 
+            bool found = false; 
+            WinRTRules winrtRules = new WinRTRules(); 
+            foreach (SyntaxTree tree in context.Compilation.SyntaxTrees) 
+            {
+                var model = context.Compilation.GetSemanticModel(tree); 
+                var nodes = tree.GetRoot().DescendantNodes();
+                
+                var classes = nodes.OfType<ClassDeclarationSyntax>().Where(winrtRules.ClassIsPublic); 
+                var interfaces = nodes.OfType<InterfaceDeclarationSyntax>().Where(winrtRules.InterfaceIsPublic);
+                var structs = nodes.OfType<StructDeclarationSyntax>();
+
+                // Used in the checking of structure fields 
+                List<string> classNames = new List<string>();

```

Comment:
Should this class name list be built earlier?  Can class names from different syntax trees be used in other syntax trees?

Reply:
I'm not sure how to make a test with multiple syntax trees, yet. Are you thinking we do one pass to gather all (qualified) class names (and interface?) and then use that here, instead of building it for each syntax tree...

Reply:
I did the idea i proposed